### PR TITLE
[dagster-dbt] add translator option to attach model files as code reference metadata

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Adds the option to a custom `DagsterDbtTranslator` to attach code reference metadata pointing to the dbt model files for each asset.

Unfortunately, this requires that we also get the `project_dir ` from the user, since that is only available at run time as part of the resource. Maybe there's a good way around this? The manifest used to have a `root_path` field, but this [was recently removed](https://github.com/dbt-labs/dbt-core/issues/6171).
 
```python
@dbt_assets(
    manifest=...,
    dagster_dbt_translator=DagsterDbtTranslator(
        settings=DagsterDbtTranslatorSettings(attach_sql_model_code_reference=True)
    ),
    project=DbtProject(...)
)
def my_dbt_assets():
  ...
```

## Test Plan

Unit tests.

### CODE REVIEW

**Comment by benpankow** [comment]
* **#21888** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21888?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py [hunk 1]
```diff
@@ -67,6 +74,7 @@ def dbt_assets(
     backfill_policy: Optional[BackfillPolicy] = None,
     op_tags: Optional[Mapping[str, Any]] = None,
     required_resource_keys: Optional[Set[str]] = None,
+    project_dir: Optional[Union[str, Path]] = None,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:77]: We should consider parameterizing this by `project: DbtProject`.

This way, we can envision a future where we deprecate the `manifest` parameter in favor of our new dbt project management scheme.

```python
@dbt_assets(project=DbtProject(project_dir=...))
def my_dbt_assets(...): ...
```

**Review by rexledesma — DISMISSED** [review]
I think this would be a good place to layer complexity on usage of `DbtProject`. This way, we can add compounding motivation for folks to shift to that abstraction.

File: python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ class DagsterDbtTranslatorSettings:
 
     enable_asset_checks: bool = True
     enable_duplicate_source_asset_keys: bool = False
+    attach_sql_model_code_reference: bool = False
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py:42]: Let's stick with the `enable_XXX` format?

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py [hunk 3]
```diff
@@ -98,6 +107,9 @@ def dbt_assets(
             are not strings will be json encoded and must meet the criteria that
             `json.loads(json.dumps(value)) == value`.
         required_resource_keys (Optional[Set[str]]): Set of required resource handles.
+        project_dir (Optional[Union[str, Path]]): The path to the dbt project directory. This directory should contain a
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:110]: update

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py [hunk 4]
```diff
@@ -67,6 +75,7 @@ def dbt_assets(
     backfill_policy: Optional[BackfillPolicy] = None,
     op_tags: Optional[Mapping[str, Any]] = None,
     required_resource_keys: Optional[Set[str]] = None,
+    project: Optional[DbtProject] = None,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:78]: This should be in a separate PR, but we should think about enforcing the fork in the road here: 

A user should either provide `manifest` or `project`, but not both.

But we should also be cognizant of this causing an instance of https://github.com/dagster-io/internal/discussions/9694.

**Review by rexledesma — COMMENTED** [review]
Could you give an example of video of this working in action? This is only for localhost right?

Also tagging in @alangenfeld as a heads up for the explicit introduction of `DbtProject` to power this feature

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py (hunk 4 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:78]: Yeah, I do think the "only one of manifest/project" is unfortunate. I'm not sure if there's a better way around it though since we can't alter the `manifest` param signature.

**Comment by alangenfeld** [comment]
cc @maximearmstrong since he's now driving moving `DbtProject` from experimental 

using `DbtProject` for this makes a ton of sense , glad it helped

File: python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py [hunk 5]
```diff
@@ -39,6 +39,7 @@ class DagsterDbtTranslatorSettings:
 
     enable_asset_checks: bool = True
     enable_duplicate_source_asset_keys: bool = False
+    enable_sql_model_code_reference: bool = False
```
[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py:42]: +1 to `enable_code_references`

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py [hunk 6]
```diff
@@ -391,12 +404,54 @@ def partitionshop_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource
...
+
+    return {
+        **existing_metadata,
+        **CodeReferencesMetadataSet(
+            code_references=CodeReferencesMetadataValue(
+                code_references=[
+                    *references,
+                    LocalFileCodeReference(
+                        file_path=str(abs_path),
+                        line_number=1,
```
[the remark this answers is not in the document]
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:441]: #21988
